# Week 2
# `https://bit.ly/2190101-02_StrList`

# Review

## expression, `math` function, `input()`, conversion

In [1]:
print(23 * 45)

1035


In [ ]:
x = input()
n = int(input())
z = float(input())

# expression
# math function
import math

# operator +, -, *, /, //, %, **, ()
# precedence of operators (order)

In [ ]:
import math

degree = int(input('enter degree: '))
radian = math.radians(degree)
sine_of_x = math.sin(radian)
print(sine_of_x)


enter degree: 30
0.49999999999999994


In [ ]:
# conversion
# str -> numbers
an_int = int('234')
a_float = float('2.34e2')
another_int = int(float('12.345e6'))

# numbers -> str
s1 = str(87)
s2 = str(0.9755)
s3 = str(3.321e-4)

## get many inputs at once

In [ ]:
# get two strings
a, b = input().split()
print(a, b)

In [ ]:
# get inputs with delimeter
hour, min, sec  = input().split('-')
print(hour, min, sec)

In [ ]:
# get integer numbers in one line
a, b, c = input().split()
a, b, c = int(a), int(b), int(c)

# better
c, d, e = [int(e) for e in input().split()]
print(c + d + e)

In [ ]:
# get floating-point numbers
x, y = [float(e) for e in input().split()]
print(x + y)

In [ ]:
# quadratic root
# ax^2 + bx + c = 0
a, b, c = [float(e) for e in input().split()]
r1 = (-b + (b**2 - 4*a*c)**0.5) / (2*a)
r2 = -b - b**2 - 4*a*c**0.5 / 2*a



## **Programing Tip:** Divide and conquer
* break complex expression into smaller expressions
* break complex task into smaller subtasks

In [ ]:
# better
t = (b**2 - 4*a*c)**0.5
bottom = 2*a
rt1 = (-b + t)/bottom
rt2 = (-b - t)/bottom

## **Exercise:** extract digit from a number

In [ ]:
a = 12345678
# want to extract hundreds from a
# 12345|6|78
#    A  ^ B
# remove A, B
a = a % 1000
print(a)
a = a // 100
print(a)

a = 12345678
a = a // 100
print(a)
print(a%10)
# want to extract millions from a
# 1|2|345678
#   ^
a = 12345678
print(a % 10000000 // 1000000)
print(a // 10000 % 10)

678
6
123456
6
2
4


# [`Share your code`](https://colab.research.google.com/drive/1su0bWuNuSwOQ25izAQbjglV0J2ZvaM5A?usp=sharing)

# Topic 2: Basic str and list operations

## `str` (string)
* `str` creation
* `len(s)`
* `str` operators



### `string creation`

In [ ]:
# string creation
#.             1
#.   01234567890123456789
# create string
s = "this is a string"
t = '12345\\c67'
u = '12   345\t67'
print(len(s), ':', s)
print(len(t), ':', t)
print(len(u), ':', u)


16 : this is a string
9 : 12345\c67
11 : 12   345	67


### `str` operators `+, *`

In [ ]:
# + operator
# concatenate
s1 = 'string1'
s2 = 'string2'
s3 = s1 + s2

In [ ]:
s1, s2, s3

('string1', 'string2', 'string1string2')

In [ ]:
# * operator
# str * integer_number
s = 'a-b'
t = 3*s
t

'a-ba-ba-b'

In [ ]:
# try multiply with 0, or negative 

## `list`
* `list` creation
* `len(d)`
* `list` operators

### `list` creation

In [ ]:
# list creation
empty_list = []
list_of_int = [1, 3, 30, -4]
list_of_float = [3.4, -4.2, 4.5e20]
weekdays = ['mon', 'tue', 'wed', 'thu', 'fri']
combination_list = list_of_float + list_of_int + weekdays
print(len(empty_list), ':', empty_list)
print(len(list_of_int), ':', list_of_int)
print(len(list_of_float), ':', list_of_float)
print(len(weekdays), ':', weekdays)
print(len(combination_list), ':', combination_list)


0 : []
4 : [1, 3, 30, -4]
3 : [3.4, -4.2, 4.5e+20]
5 : ['mon', 'tue', 'wed', 'thu', 'fri']
12 : [3.4, -4.2, 4.5e+20, 1, 3, 30, -4, 'mon', 'tue', 'wed', 'thu', 'fri']


### `list` operator `+, *`

In [ ]:
# list operator
x = [2, 3]
y = [5, 6]
print(x + y, y + x)

[2, 3, 5, 6] [5, 6, 2, 3]


In [ ]:
# * operator
x = [2, 3]
print(2*x, x*3)

[2, 3, 2, 3] [2, 3, 2, 3, 2, 3]


#### What is the result when multply a list or str with 0 or negative number?

In [ ]:
# try it

### **Exercise 2-1**: Create variable x, which is a list of integer. x has 300 slots. The first 100 slots store 0, the second 100 slots store 1, the rest store 2.

In [ ]:
x = 100*[0] + 100*[1] + 100*[2]

## `list` from `str`, `str` from `list`(of `str`)

In [ ]:
s = 'abcde  x\\l xxx'
list_from_str = list(s)
print(s)
print(list_from_str)

In [ ]:
s = 'Today is 16 Aug 2022.'
x = list(s)
print(x)
print(s.split())

In [ ]:
student_id = '6431234521'
x = list(student_id)
print(x)

In [ ]:
weekdays = ['mon', 'tue', 'wed', 'thu', 'fri']
str_from_list = '*-*'.join(weekdays) 
print(str_from_list)

## access elements in `str` and `list`

`n` is length of a `str` or `list`
* `0 --> n-1` left to right
* `-1 --> -n` right to left

In [ ]:
s = 'abcdefghij' 
# --> 0, 1, 2, ..., n-1
#     -n, -n+1, -n+2, ..., -2, -1 <--
print(s[0])
print(s[2])
print(len(s))
print(s[-1])
print(s[-8])
print(s[-9])
print(s[-10])
print(s[-11])

In [ ]:
d = [1, 3, 30, -4]
print(len(d), ':', d)
print(d[0])
print(d[2])
print(d[-1])
print(d)

## changing 1 element inside a `str` or `list`

In [ ]:
s = '12345'
s[0] = '0' # ERROR, str cannot be modified

In [ ]:
s = '12345'
t = list(s)
print(t)
t[3] = "xx"
t[0] = 99
print(t)
print(t[-1])
t[-1] = int(t[-1])
print(t)

### **Exercise 2-2**: 
1. Read a string from input and store it in `s`.
2. Create a variable `t` (t must be a string) that store the left most character and the rightmost character of s
3. Store the middle character of `s` in variable `u`. If `s = "01234"` the middle character is `"2"`. If `s = "012345"`, the middle character is `"2"`.

In [27]:
import math

s = input()
t = s[0] + s[-1]
middleIdx = len(s) // 2
u = s[middleIdx]

print(t, u)

04 1


## slicing `x[start:stop:step]`

* positive step 
** `start` must be left of `stop`

* negative step 
** `start` must be right of `stop`

otherwise, empty result

In [ ]:
#----0.        1.        2.        3----'
#----0123456789012345678901234567890----'
x = 'abcdefghijklmnopqrstuvwxyz'
#x[0:len(x):1], x[0:len(x)], x[:len(x)], x[0:], x[:]
#x[5:], x[5:len(x):1]
#x[:5], x[0:5:1]
#x[5::2]
x[1::2]

In [ ]:
# reverse a string
print(x[::-1]) 
print(x[-1:-len(x):-1])
# how to correct the second one?

In [ ]:
# only first 2 characters
# only last 2 characters
x[:2], x[-2:]

In [ ]:
# remove first 2 characters or remove last 2 characters
x[2:], x[:-2]

### Use list slicing to modify list content


In [47]:
x = list('1234567890')
print(x)
#x[:3] = 0
#x[:3] = [0]
print(x[:3])
x[:3] = 3*[0]
print(x)

['1', '2', '3', '4', '5', '6', '7', '8', '9', '0']
['1', '2', '3']
[0, 0, 0, '4', '5', '6', '7', '8', '9', '0']


### **Exercise 2-3:** How can we change the last character of a string to a new character?

Example: `abcde` ==> `abcd*`

In [ ]:
s = 'abcde' # => abcd*



### **Exercise 2-4:** `v1` and `v2` are given lists, each with 3 slots. Each of them represents a vector. 

Show the dot product of `v1` and `v2`.

In [ ]:
v1 = [2,3,4]
v2 = [4,5,6]


### **Exercise 2-5:** Break a string `s` into 2 parts, first half and second half.

In [ ]:
s = '1234abcd' # first -> '1234', second -> 'abcd'


### **Exercise 2-6:** Using list slicing to remove element(s) from a list

In [65]:
x = list('1234567')
print(x, id(x))
# use list slicing to remove 1 or more elements
x[0:1] = []

print(x, id(x))

['1', '2', '3', '4', '5', '6', '7'] 1813226459520
['2', '3', '4', '5', '6', '7'] 1813226459520


### **Exercise 2-7:** Change odd locations to zeros.

In [56]:
x = list('1234567890') # => 1,0,3,0,5,0,7,0,9,0
# use list slicing to change odd locations to zeros

x[1::2] = 5 * [0]

print(x)

['1', 0, '3', 0, '5', 0, '7', 0, '9', 0]


### **Exercise 2-8:** Merge two lists

In [67]:
x = list('12345')
y = list('abcde')
print(x, y)
# use list slicing to merge x and y 
# 1, a, 2, b, 3, c, 4, d, 5, e
z = len(x)*2*[0]
z[::2] = x
z[1::2] = y
print(z)


['1', '2', '3', '4', '5'] ['a', 'b', 'c', 'd', 'e']


['1', 'a', '2', 'b', '3', 'c', '4', 'd', '5', 'e']

## **Example:** using `list` to convert numbers to `str`

In [ ]:
# convert number to word
number_text = ['zero', 'one', 'two', 'three', 'four', \
               'five', 'six', 'seven', 'eight', 'nine']
print(number_text[7])

month = ['January', 'February', 'March', 'April', 'May', 'June', \
         'July', 'August', 'September', 'October', 'November', 'December']
n = 3
offset = 1  # starting number
print(month[n - offset])

## Convert a string to list using `s.split(), s.split(separator)`

In [ ]:
x = 'ab:de'
#x = 'ab::de: fg'
#x = input('> ')
print(list(x))
print(x.split())
print(x.split(':'))

### **Exercise 2-9:** 
1. get keyboard input in format `day/month/year`
2. show result in format `month-name day, year`

For example, `16/8/2022` should be displayed as `August 16, 2022`

In [ ]:
# 26/8/2563 --> d:26, m:8, y:2563
# d/m/y --> m/d/y
x = input('enter d/m/y: ')


## Advance `str` output

In [ ]:
# 9 -> 009
# 10 -> 010
# 123 -> 123
n = 3
number = 1
s = str(number)
t = '0'*n + s
print(s, t)
# how to print only 3 digits?

1 0001


In [ ]:
import random

ruler = '---------1---------2---------3---------4---------5' 
line_len = len(ruler)
print(ruler)

topics = ['Preface', 'Introduction', 'Chapter 1', 'Expression, Variable', 'How to use String, List', 'Summary']

n = random.randint(0, len(topics) - 1) # random integer x, 0 <= x <= len(topics) - 1
topic = topics[n]

page = random.randint(1,10000)

print(topic, page)

dots_len = line_len - len(topic) - len(str(page))
dots = '.' * dots_len

print(topic + dots + str(page))

page = 5295456
print(page)
text = '0'*10 + str(page)
print(text)
print(text[-10:])

n_zeros = 10 - len(str(page))
text = n_zeros * '0' + str(page)
print(text)

text = ' '*20 + str(page)
print(text[-20:])

---------1---------2---------3---------4---------5
Chapter 1 3429
Chapter 1.....................................3429
5295456
00000000005295456
0005295456
0005295456
             5295456


## Functions (How to call a function?)


```
def function_name(parameters):
  statement 1
  statement 2
   ...
  statement n
```


In [ ]:
def main():
    print('hello')
    f()
    f()

def f():
    print('function f()')


main()



hello
function f()
function f()


In [ ]:
x = int(input())
y = int(input())
z = x**2 + y**3 + 2*(x+y)
print(z)

20
3
473


In [ ]:
def fz(x, y):
    z = x**2 + y**3 + 2*(x+y)
    return z

def main():
    x = int(input())
    y = int(input())
    z = fz(x, y)
    print(z)

main()

In [ ]:
import math
print(3.1**4.3, math.pow(3.1, 4.3))

129.67470942124882 129.67470942124882


## Grader `02_StrList`


### CitizenID

In [ ]:
# get each digit 
# - check
# convert to integers
# use formula to calculate

### Arabic number

In [ ]:
# Arabic numerals
# create a list
# get data from list
# convert to str, see example

### US Date

In [ ]:
# US Date
# get keyboard input -> date, month, year
# convert month -> month name using list, see Arabic number as example
# combine result and display


### Prefix

In [ ]:
# prefix with 0
# get input and find to length of it
# find number of digit to be displayed, k (without using if)
#   hint: using math function
# add '0's in front of it
# show k digits 

### Vector

In [ ]:
# ผลบวกเวคเตอร์
# รับข้อมูลมาเป็นสตริง 
#   '[จำนวน, จำนวน, จำนวน]' 
#    => 'จำนวน, จำนวน, จำนวน'
#.   => ['จำนวน' 'จำนวน' จำนวน']
#  แปลงให้เป็น จำนวน จำนวน จำนวน
# แปลงให้เป็นลิสต์ของจำนวน [u0, u1, u2], [v0, v1, v2]
# คำนวณ w = u + v -> [u0+v0, u1+v1, u2+v2]
# แสดงผลการคำนวณ
# [P P [PP] [PP] [PP]], [PP] ถึงจะถือว่าถูก [P-], [-P], [--]

## Decimal2Fraction

In [ ]:
# 